src/docs/asciidoc/Optimizing-Serialization.asciidoc

```diff
@@ -0,0 +1,270 @@
+with no partitioning or parallel deployment, is at most 100,000 messages/sec.
+This does not account for network latency (transporting the serialized bytes),
+overhead in the messaging middleware, or computations performed by the modules.
+If your application requires this level of throughput then serialization is a
+potential bottleneck. However, if your throughput requirements are not so
+demanding, serialization performance becomes less of an issue. In this
+hypothetical scenario, At a throughput of 10,000 messages/second, 10
+microseconds is only 10% of the maximum time required to process the message.
+However if you trying to stream large objects at this rate, serialization time,
+along with transport and processing time, will be significantly higher.
+
+=== Serialization in XD
+
+NOTE: This content applies to Spring XD 1.2.x. In Spring XD 2.0, these
+components work in a similar way but will be packaged and configured differently.
+
+In Spring XD 1.2.x, MessageBus implementations inject a bean of type
+http://docs.spring.io/spring-xd/docs/current/api/org/springframework/xd/dirt/integration/bus/serializer/MultiTypeCodec.html[MultiTypeCodec],
+where the term `Codec` refers to an component providing methods to serialize and
+deserialize objects. `MultiType` means that the component can handle any object. Spring
+XD is internally configured with a
+http://docs.spring.io/spring-xd/docs/current/api/org/springframework/xd/dirt/integration/bus/serializer/kryo/PojoCodec.html[PojoCodec]
+that delegates to Kryo and provides hooks to register custom Kryo serializers for
+specific types.
+
+=== Customizing Kryo
+
+By default, Kryo uses delegates unknown Java types to its `FieldSerializer`.
+Kryo also registers default serializers for each primitive type along with
+`String`, `Collection` and `Map` serializers. `FieldSerializer` uses reflection
+to navigate the object graph. A more efficient approach is to implement a custom
+serializer that is aware of the object's structure and can directly serialize
+selected primitive fields:
+
+[source,java]
+----
+public class AddressSerializer extends Serializer<Address> {
+	@Override
+	public void write(Kryo kryo, Output output, Address address) {
+		output.writeString(address.getStreet());
+		output.writeString(address.getCity());
+		output.writeString(address.getCountry());
+	}
+
+	@Override
+	public Address read(Kryo kryo, Input input, Class<Address> type) {
+		return new Address(input.readString(),input.readString(),input.readString());
+	}
+}
+----
+
+The `Serializer` interface exposes `Kryo`, `Input`, and `Output` which provide
+complete control over which fields are included and other internal settings as
+described in the https://github.com/EsotericSoftware/kryo[documentation].
+
+
+==== Disabling References
+
+A simple setting may be applied to boost performance if your payload types do
+not contain any cyclical references. Kryo needs to do some extra work in this
```

Comment:
Reworded to make this clearer.
If payload contains cyclical references Kryo needs to do some extra work in this case and the https://github.com/EsotericSoftware/kryo#references[references]
+setting is enabled by default.